go/vt/grpcclient/client_auth_static.go

```diff
@@ -52,17 +56,26 @@ func (c *StaticAuthClientCreds) RequireTransportSecurity() bool {
 	return false
 }
 
-// AppendStaticAuth optionally appends static auth credentials if provided.
-func AppendStaticAuth(opts []grpc.DialOption) ([]grpc.DialOption, error) {
-	if credsFile == "" {
-		return opts, nil
-	}
+// loadStaticAuthCredsFromFile loads static auth credentials from a file.
+func loadStaticAuthCredsFromFile(credsFile string) (*StaticAuthClientCreds, error) {
 	data, err := os.ReadFile(credsFile)
 	if err != nil {
 		return nil, err
 	}
 	clientCreds := &StaticAuthClientCreds{}
 	err = json.Unmarshal(data, clientCreds)
+	return clientCreds, err
+}
+
+// AppendStaticAuth optionally appends static auth credentials if provided.
+func AppendStaticAuth(opts []grpc.DialOption) ([]grpc.DialOption, error) {
+	if credsFile == "" {
+		return opts, nil
+	}
+	var err error
+	credsFileOnce.Do(func() {
```

Comment:
I suppose this could happen in `init()` or elsewhere, although here we are able to capture an `error` 🤔

Reply:
you can also store the error if you run in an init:

```go
func init() {
    if credsFile == "" {
        return opts, nil
    }
    
    clientCreds, clientCredsErr = loadStaticAuthCredsFromFile(credsFile)
}
```

and then check that package-global whenever we use the creds to fail immediately

Reply:
also worth noting this is technically a breaking change, as it's possible someone was (ab)using this behavior to hot swap client creds (rewrite the existing file in place, get new creds on next connection)

Reply:
One way of mitigating that would be to reload the file on a SIGHUP. We do that with other credential files.

Reply:
Yes, I pondered something similar here:
> A consequence of this change is a process reload is required to change the file, whereas today you could change the file and it would be re-loaded (sadly by N x goroutines) on each loop of [the topology watcher .Start(...) ticker](https://github.com/slackhq/vitess/blob/main/go/vt/discovery/topology_watcher.go#L126), which to me isn't a dealbreaker (I was surprised it's reloaded actually), but I'm curious what everyone thought? I'm happy to adjust this to support SIGHUP and I'm curious if there are helpers for that sort of thing

I feel like reload-ability was never promised for this file but having `SIGHUP` reload functionality would be useful 👍. I suppose that would still be a breaking change for those changing the file without any other actions like @ajm188 mentioned - there's probably no way to avoid that aside from using maybe `fsnotify`

Reply:
I think this is fine since it was never advertised as a feature (just an accidental implementation side-effect), but it might be worth bringing it up at the monthly meeting if you are coming tomorrow!

Adding SIGHUP would be a nice bonus, too :)

Reply:
I agree that this should be fine. I can't imagine that anyone is changing their "static" creds every second and depending on file reload. If they are they will just need to adjust to the new way. Unblocking concurrency and performance is much more important.

Reply:
@ajm188 just an update: moving this logic to `init()` didn't quite work, so I've moved it out of `init()`

Putting the logic in `init()` caused the loading + watcher of the file to start before the CLI-flag library actually sets `credsFile`, so they always see `""` as the file name